Implement betaincinv and gammainc[c]inv functions

amyoshino opened the pull request:
<!-- !! Thank your for opening a PR !! -->

### Motivation for these changes
Closes https://github.com/pymc-devs/pytensor/issues/413

Implement `betaincinv` and `gammaincinv` functions.
The implementation of these functions will allow us to implement quantile/ICDF functions for distributions like Beta, Gamma, ChiSquared and StudentT distributions, being fundamental to the issue: https://github.com/pymc-devs/pymc/pull/6845

### Implementation details
Following other scalar.math functions implementations, the code is extended to call the functions from `scipy.special`.
Tests were added to check if it matches scipy implementation.


### Checklist
+ [ ] Explain motivation and implementation 👆
+ [ ] Make sure that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html).
+ [ ] Link relevant issues, preferably in [nice commit messages](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
+ [ ] The commits correspond to [_relevant logical changes_](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes). Note that if they don't, we will [rewrite/rebase/squash the git history](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History#_rewriting_history) before merging.
+ [ ] Are the changes covered by tests and docstrings?
+ [ ] Fill out the short summary sections 👇


## Major / Breaking Changes
- ...

## New features
- ...

## Bugfixes
- ...

## Documentation
- ...

## Maintenance
- ...

amyoshino commented:
I finally was able to get my hands back to it, apologies for resuming it very late. 

I have only implemented the function "impl" so far, gradients were not implemented and not sure if it is needed in the case of those inverse regularized functions. 

Let me know what I am missing anything.

ricardoV94 commented on tests/scalar/test_math.py:
```diff
@@ -59,6 +61,14 @@ def test_gammaincc_nan_c():
     assert np.isnan(test_func(-1, -1))
 
 
+def test_gammaincinv_python():
```
Usually we don't test the Scalar version unless it's something fancy. Here we're just calling scipy.

Instead we test the Elemwise version in pytensor.tensor which is usually what users see.

ricardoV94 commented on pytensor/scalar/math.py:
```diff
@@ -684,6 +684,27 @@ def __hash__(self):
 gammaincc = GammaIncC(upgrade_to_float, name="gammaincc")
 
 
+class GammaIncInv(BinaryScalarOp):
```
Wolfram suggests the derivative wrt to `s` (I don't know if that's k or x) is relatively simple, so it would be nice to add

See scope>differentiation 

https://reference.wolfram.com/language/ref/InverseGammaRegularized.html

Would also be nice to check if there are simple formula for the derivatives of the invbetainc

amyoshino replied:
Thanks for the pointer!! I feel bad for not checking it before, will definitely work on it

ricardoV94 replied:
No worries, that's how we all start

ricardoV94 reviewed (commented):
Thanks for picking it up!

ricardoV94 commented:
@amyoshino here is a more typical PR to add new scalar/elemwise Ops that you can use as a template: https://github.com/pymc-devs/pytensor/commit/2dc759127547d4b836629e927b7af24f840cc91c

ricardoV94 commented:
For JAX dispatching, it seems like these exist in tensorflow-probability:

https://www.tensorflow.org/probability/api_docs/python/tfp/substrates/jax/math/betaincinv
https://www.tensorflow.org/probability/api_docs/python/tfp/substrates/jax/math/igammainv

We can do something like https://github.com/pymc-devs/pytensor/pull/403/files to provide them in the JAX backend

amyoshino commented on pytensor/scalar/math.py:
```diff
@@ -1568,6 +1589,23 @@ def inner_loop(
+    Inverse of the regularized incomplete beta function.
+    """
+
+    nfunc_spec = ("scipy.special.betaincinv", 1, 1)
```
Thanks!!! I got that wrong, my bad!

amyoshino commented:
> [2dc7591](https://github.com/pymc-devs/pytensor/commit/2dc759127547d4b836629e927b7af24f840cc91c)

I see I am missing quite a few things, thanks for the guidance!

amyoshino commented:
I have added the first derivatives for `betaincinv`, `gammaincinv` (lower) and `gammainccinv` (upper) and tried to set up some Elemwise tests. 

I have not set up the JAX dispatching yet, but plan to do so in the next PR.

Thanks for all your guidance, and I am hoping to improve it soon!

amyoshino commented:
@ricardoV94 Phew, what a journey to get all tests good for this PR. I believe now it is all good and it is ready for review. 
Again, sorry for the slow pace in this issue. 

Let me take this comment to summarize all the steps and assumptions used in this PR, I hope it helps the reviewing process.

Overall I implemented the wrapper of  `scipy.special.gammaincinv`, `scipy.special.gammainccinv` and `scipy.special.betaincinv` with:
- the first derivatives implementation under the `grad` function, with respect to arguments that I was able to find closed form solutions. The ones I wasn't able I added `grad_not_implemented`
- Include JAX dispatching when available
- Write tests to cover outputs and broadcasting

In order to write the derivative for `scipy.special.betaincinv` w.r.t. `x`, I needed to add `scipy.special.beta` (beta function), and with it, add its derivatives wrt `a` and `b` and required tests for it. I wasn't able to find the JAX version for this function, so I haven't added it as well.

This PR will help the issue https://github.com/pymc-devs/pymc/pull/6845, which will benefit from the implemented functions for the icdf formulas.

ricardoV94 commented:
@amyoshino sorry I still didn't have time to review, just this comment here caught my attention:

>  I needed to add scipy.special.beta (beta function), and with it, add its derivatives wrt a and b and required tests for it. I wasn't able to find the JAX version for this function, so I haven't added it as well.

Instead of implementing a new Op, let's use the mathematical definition instead:
```python
def betaln(x, y):
    return gammaln(x) + gammaln(y) - gammaln(x + y)
```

This saves us from having to maintain another Op. We can also add a helper in `pytensor.math` just for that

amyoshino commented:
@ricardoV94 enjoy your holiday time, no worries about it. I delayed the progress of this PR for so long, it would be really nonsense on my end to expect a quick response in this 🥲 

Nice suggestions! I will make the changes in the code shortly, and for safety, I'll try to find out if the proposed solution is not unstable in any ways. I will keep you posted on the progress.

amyoshino commented:
I investigated the stability of using the proposed form of the `beta function`, and here are my findings so far. 

Since the connection between beta and gamma function is valid only when `x > 0 `and `y > 0` ([reference](https://www.jstor.org/stable/2306926)), as expected, the approximation works well under this condition, and implementing it with `gammln` is preferred over `gamma` because of numerical instability on large values:

### Unstable when arguments have "high" values:
<img width="725" alt="image" src="https://github.com/pymc-devs/pytensor/assets/22087841/6d412324-2e36-445b-99ea-c9821ecb48e0">

### Approximation is great (only one tiny difference found in a relatively large range of arguments)
<img width="743" alt="image" src="https://github.com/pymc-devs/pytensor/assets/22087841/683aba73-13bb-457e-9503-156533081bc6">

## When x < 0. and/or y < 0
The thing starts to go a bit wrong if we want to keep consistency in the results between this implementation and the SciPy's Beta function when `x < 0` and/or `y < 0`.
When checking for these cases, then we find some inconsistencies. Just to illustrate:

### Results can change sign depending on the argument value
<img width="750" alt="image" src="https://github.com/pymc-devs/pytensor/assets/22087841/690f9ddf-8957-4b1a-85b0-3f230a6c21da">

### And also can have a large difference when values are high (only when one or more arguments are < 0):
<img width="843" alt="image" src="https://github.com/pymc-devs/pytensor/assets/22087841/d7064405-9e7a-4c63-b0b5-525f54380bac">
<br></br>
Given this information, should we go ahead and use your suggestion? I have already worked on the changes, whatever path you want to take is good for me. Let me know and can push it at anytime. 😄

ricardoV94 commented:
Aren't negative values some special analytic extension? Or is it standard to support it?

ricardoV94 commented:
> Given this information, should we go ahead and use your suggestion?

My gut feeling is we should use exp(beta_ln) in the scalar grads and offer beta and betaln to user in tensor.math.

Let's see if the gradients don't blowup for some reasonable values of the original functions. you can test locally with verify_grad

amyoshino commented:
Thanks for the suggestions and pointers to where to look at! I will evaluate all the mentioned points.

> We should think about the stability and range in the context that first motivated these Ops: use in the gradients of this PR.
> 
> Do the gradients remain stable in a reasonable parameter space of the original functions?
 
Good call, I will check it and be mindful about the objective of the PR.

> Aren't negative values some special analytic extension? Or is it standard to support it?

I will research more about it. Since the function is well defined only for positive values, I imagine it is kind of a special use, but let me find out in which circumstances negative values are used before stating anything. 

> Also for assessing closeness it's common to look not only at absolute error but also relative error. Those 1e+16 may not be so unreasonable if they correspond to a small relative error

Most of errors are change in sigs, so error is close to zero. But in few cases errors are large, but only happens when values are high as well, so I will evaluate if this is really a concern or if large errors happen only when the functions are itself unusable. 

> Let's see if the gradients don't blowup for some reasonable values of the original functions. you can test locally with verify_grad

Thank you so much for the hint! I will need some time to check the gradients and be more familiarized with the trace to the gradient computation, having a tip on where to start is very valuable 😄

amyoshino commented:
@ricardoV94, after checking the review comments, here is my suggestion:

>Do the gradients remain stable in a reasonable parameter space of the original functions?

Given the context of the PR, the use of beta function for the inverse of beta distribution gradient computation will not present us negative values as arguments (as parameters of the `beta distribution` must be `Positive and real-valued`). That said, I believe we should go with the approximation you suggested. 

Moreover, I could not find use cases for negative-valued arguments for the beta function. So, if any use case comes up in the future, we can revise the user-facing helper in `tensor.special` and implement the `Op` for the `beta function`.

> Let's see if the gradients don't blowup for some reasonable values of the original functions. you can test locally with verify_grad

For safety and consistency with previously implemented `betainc_grad()` function, I used the `betaln` approach. This is because having alpha or beta with a value of 180 (taking the blown-up example) is not unlikely.

> We can also add a user-facing helper in tensor.math (the scalar version used in the grads wouldn't be user facing)

I have done that as well! Thanks for suggesting it. 

I just pushed the latest code, I hope everything is good, but if not, do not hesitate in pointing out improvements, I am never tired of improving things and keeping the codebase as neat as possible in a single pass. 😄

ricardoV94 commented on pytensor/tensor/special.py:
```diff
@@ -752,9 +752,20 @@ def factorial(n):
+    return gammaln(a) + gammaln(b) - gammaln(a + b)
+
+
+__all__ = ["softmax", "log_softmax", "poch", "factorial", "beta", "betaln"]
```
If you add a comma after the last entry, the pre-commit will default to multi-line, which is better for git changes

```suggestion
__all__ = ["softmax", "log_softmax", "poch", "factorial", "beta", "betaln",]
```

amyoshino replied:
Thanks for the suggestion! I didn't know about it, and I felt weird that pre-commit had arranged this line in this way, but didn't want to overwrite the pre-commit changes by myself. 

I tried to accept the suggestion through the website UI but it apparently didn't run pre-commit automatically, so I ran in my machine and pushed it again.

ricardoV94 commented:
@amyoshino this looks great, I left a tiny tiny comment above but it looks good to merge!

amyoshino commented:
@ricardoV94 As always, thanks for guididing me through this PR. Learned a lot about the codebase and tests with this one.

ricardoV94 commented:
My pleasure, and thanks for taking this one on. It will be really useful for the icdf methods!
